rust/ql/lib/codeql/rust/dataflow/internal/DataFlowImpl.qll

```diff
@@ -0,0 +1,324 @@
+  final class NaNode extends Node {
+    NaNode() { none() }
+
+    override string toString() { result = "N/A" }
+
+    override Location getLocation() { none() }
+  }
+
+  /**
+   * The value of a parameter at function entry, viewed as a node in a data
+   * flow graph.
+   */
+  final class ParameterNode extends Node {
+    Param param;
+
+    ParameterNode() { this = TSourceParameterNode(param) }
+
+    override Location getLocation() { result = param.getLocation() }
+
+    override string toString() { result = param.toString() }
+  }
+
+  final class ArgumentNode = NaNode;
+
+  final class ReturnNode extends NaNode {
+    ReturnKind getKind() { none() }
+  }
+
+  final class OutNode = NaNode;
+
+  /**
+   * A node associated with an object after an operation that might have
+   * changed its state.
+   *
+   * This can be either the argument to a callable after the callable returns
+   * (which might have mutated the argument), or the qualifier of a field after
+   * an update to the field.
+   *
+   * Nodes corresponding to AST elements, for example `ExprNode`, usually refer
+   * to the value before the update.
+   */
+  final class PostUpdateNode extends Node::NaNode {
+    /** Gets the node before the state update. */
+    Node getPreUpdateNode() { none() }
+  }
+
+  final class CastNode = NaNode;
+}
+
+module RustDataFlow implements InputSig<Location> {
+  /**
+   * An element, viewed as a node in a data flow graph. Either an expression
+   * (`ExprNode`) or a parameter (`ParameterNode`).
+   */
+  final class Node = Node::Node;
+
+  final class ParameterNode = Node::ParameterNode;
+
+  final class ArgumentNode = Node::ArgumentNode;
+
+  final class ReturnNode = Node::ReturnNode;
+
+  final class OutNode = Node::OutNode;
+
+  final class PostUpdateNode = Node::PostUpdateNode;
+
+  final class CastNode = Node::NaNode;
+
+  predicate isParameterNode(ParameterNode p, DataFlowCallable c, ParameterPosition pos) { none() }
+
+  predicate isArgumentNode(ArgumentNode n, DataFlowCall call, ArgumentPosition pos) { none() }
+
+  DataFlowCallable nodeGetEnclosingCallable(Node node) { none() }
+
+  DataFlowType getNodeType(Node node) { none() }
+
+  predicate nodeIsHidden(Node node) { none() }
+
+  class DataFlowExpr = Void;
+
+  /** Gets the node corresponding to `e`. */
+  Node exprNode(DataFlowExpr e) { none() }
+
+  final class DataFlowCall extends TNormalCall {
+    private CallExpr c;
+
+    DataFlowCall() { this = TNormalCall(c) }
+
+    DataFlowCallable getEnclosingCallable() { none() }
+
+    string toString() { result = c.toString() }
+
+    Location getLocation() { result = c.getLocation() }
+  }
+
+  final class DataFlowCallable = CfgScope;
+
+  final class ReturnKind = Void;
+
+  /** Gets a viable implementation of the target of the given `Call`. */
+  DataFlowCallable viableCallable(DataFlowCall c) { none() }
+
+  OutNode getAnOutNode(DataFlowCall call, ReturnKind kind) { none() }
+
+  final class DataFlowType = Void;
```

Comment:
I think nodes are required to have a (dummy) type, so this should be `Unit` and then `compatiableTypes` should be `any()`.